File: frontend/src/services/mfaService.js
import axios from 'axios';

const API_URL = 'http://localhost:8081/api/auth/mfa';

/**
 * MFA Service - API calls for Multi-Factor Authentication
 * 
 * @author Bob
 */
class MfaService {
    /**
     * Setup MFA - Get QR code and secret
     */
    async setupMfa() {
        const token = localStorage.getItem('token');
        
        if (!token) {
            throw new Error('No authentication token found');
        }

        try {
            const response = await axios.post(
                `${API_URL}/setup`,
                {},
                {
                    headers: {
                        'Authorization': `Bearer ${token}`
                    }
                }
            );
            return response.data;
        } catch (error) {
            throw error.response?.data?.error || 'Failed to setup MFA';
        }
    }

    /**
     * Enable MFA with verification code
     */
    async enableMfa(secret, code) {
        const token = localStorage.getItem('token');
        
        if (!token) {
            throw new Error('No authentication token found');
        }

        try {
            const response = await axios.post(
                `${API_URL}/enable`,
                { secret, code },
                {
                    headers: {
                        'Authorization': `Bearer ${token}`,
                        'Content-Type': 'application/json'
                    }
                }
            );
            return response.data;
        } catch (error) {
            throw error.response?.data?.error || 'Failed to enable MFA';
        }
    }

    /**
     * Disable MFA
     */
    async disableMfa(code) {
        const token = localStorage.getItem('token');
        
        if (!token) {
            throw new Error('No authentication token found');
        }

        try {
            const response = await axios.post(
                `${API_URL}/disable`,
                { code },
                {
                    headers: {
                        'Authorization': `Bearer ${token}`,
                        'Content-Type': 'application/json'
                    }
                }
            );
            return response.data;
        } catch (error) {
            throw error.response?.data?.error || 'Failed to disable MFA';
        }
    }

    /**
     * Check MFA status
     */
    async getMfaStatus() {
        const token = localStorage.getItem('token');
        
        if (!token) {
            throw new Error('No authentication token found');
        }

        try {
            const response = await axios.get(
                `${API_URL}/status`,
                {
                    headers: {
                        'Authorization': `Bearer ${token}`
                    }
                }
            );
            return response.data;
        } catch (error) {
            throw error.response?.data?.error || 'Failed to get MFA status';
        }
    }

    /**
     * Login with MFA code
     */
    async loginWithMfa(username, password, mfaCode) {
        try {
            const response = await axios.post(
                'http://localhost:8081/api/auth/login/mfa',
                {
                    username,
                    password,
                    mfaCode: parseInt(mfaCode)
                },
                {
                    headers: {
                        'Content-Type': 'application/json'
                    }
                }
            );

            if (response.data.token) {
                localStorage.setItem('token', response.data.token);
                localStorage.setItem('user', JSON.stringify({
                    username: response.data.username,
                    email: response.data.email
                }));
            }

            return response.data;
        } catch (error) {
            throw error.response?.data?.error || 'MFA login failed';
        }
    }
}

export default new MfaService();

// Made with Bob

File: backend/src/main/java/com/login/service/MfaService.java
package com.login.service;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.client.j2se.MatrixToImageWriter;
import com.google.zxing.common.BitMatrix;
import com.login.model.User;
import com.login.repository.UserRepository;
import com.warrenstrange.googleauth.GoogleAuthenticator;
import com.warrenstrange.googleauth.GoogleAuthenticatorKey;
import com.warrenstrange.googleauth.GoogleAuthenticatorQRGenerator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Base64;

/**
 * Multi-Factor Authentication Service
 * Handles TOTP (Time-based One-Time Password) generation and verification
 * 
 * @author Bob
 */
@Service
public class MfaService {

    private static final String ISSUER = "LoginMicroservice";
    private static final int QR_CODE_SIZE = 200;

    @Autowired
    private UserRepository userRepository;

    private final GoogleAuthenticator googleAuthenticator;

    public MfaService() {
        this.googleAuthenticator = new GoogleAuthenticator();
    }

    /**
     * Generate MFA secret for a user
     * 
     * @param user The user
     * @return The generated secret key
     */
    public String generateSecret(User user) {
        GoogleAuthenticatorKey key = googleAuthenticator.createCredentials();
        return key.getKey();
    }

    /**
     * Generate QR code URL for Google Authenticator
     * 
     * @param user The user
     * @param secret The MFA secret
     * @return QR code URL
     */
    public String generateQrCodeUrl(User user, String secret) {
        return GoogleAuthenticatorQRGenerator.getOtpAuthURL(
            ISSUER,
            user.getUsername(),
            new GoogleAuthenticatorKey.Builder(secret).build()
        );
    }

    /**
     * Generate QR code image as Base64 string
     * 
     * @param qrCodeUrl The QR code URL
     * @return Base64 encoded QR code image
     */
    public String generateQrCodeImage(String qrCodeUrl) {
        try {
            BitMatrix bitMatrix = new MultiFormatWriter().encode(
                qrCodeUrl,
                BarcodeFormat.QR_CODE,
                QR_CODE_SIZE,
                QR_CODE_SIZE
            );

            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            MatrixToImageWriter.writeToStream(bitMatrix, "PNG", outputStream);
            byte[] qrCodeBytes = outputStream.toByteArray();
            
            return "data:image/png;base64," + Base64.getEncoder().encodeToString(qrCodeBytes);
        } catch (WriterException | IOException e) {
            throw new RuntimeException("Failed to generate QR code", e);
        }
    }

    /**
     * Verify TOTP code
     * 
     * @param secret The MFA secret
     * @param code The TOTP code to verify
     * @return true if code is valid, false otherwise
     */
    public boolean verifyCode(String secret, int code) {
        return googleAuthenticator.authorize(secret, code);
    }

    /**
     * Enable MFA for a user
     * 
     * @param user The user
     * @param secret The MFA secret
     * @param verificationCode The verification code to confirm setup
     * @return true if MFA was enabled successfully
     */
    @Transactional
    public boolean enableMfa(User user, String secret, int verificationCode) {
        // Verify the code before enabling MFA
        if (!verifyCode(secret, verificationCode)) {
            throw new RuntimeException("Invalid verification code");
        }

        user.setMfaSecret(secret);
        user.setMfaEnabled(true);
        userRepository.save(user);
        
        return true;
    }

    /**
     * Disable MFA for a user
     * 
     * @param user The user
     * @param verificationCode The verification code to confirm disable
     * @return true if MFA was disabled successfully
     */
    @Transactional
    public boolean disableMfa(User user, int verificationCode) {
        if (!user.isMfaEnabled()) {
            throw new RuntimeException("MFA is not enabled for this user");
        }

        // Verify the code before disabling MFA
        if (!verifyCode(user.getMfaSecret(), verificationCode)) {
            throw new RuntimeException("Invalid verification code");
        }

        user.setMfaSecret(null);
        user.setMfaEnabled(false);
        userRepository.save(user);
        
        return true;
    }

    /**
     * Check if user has MFA enabled
     * 
     * @param user The user
     * @return true if MFA is enabled
     */
    public boolean isMfaEnabled(User user) {
        return user.isMfaEnabled() && user.getMfaSecret() != null;
    }

    /**
     * Validate MFA code for login
     * 
     * @param user The user
     * @param code The TOTP code
     * @return true if code is valid
     */
    public boolean validateMfaCode(User user, int code) {
        if (!isMfaEnabled(user)) {
            throw new RuntimeException("MFA is not enabled for this user");
        }

        return verifyCode(user.getMfaSecret(), code);
    }
}

// Made with Bob
